### PR TITLE
Issue 1433: Expose batch flushAsync() and flush() methods in Producer

### PR DESCRIPTION
**Pull request by sijie**
*Motivation*

Currently when you're using batches server will wait until timeout to send messages.

To archive the maximum theoretical performance I suggest to expose flushAsync method (and flush() for completeness) of the Producer class.

That will force publisher to flush all batched messages for faster delivery (faster than waiting until batch timeout).

*Changes*

- Introduce `flush` and `flushAsync` in Producer interface
- rename existing `flush` method to `triggerFlush`
- flush is waiting for the completion of last write
7345ae1

### CODE REVIEW

File: pulsar-client/src/main/java/org/apache/pulsar/client/api/Producer.java [hunk 1]
```diff
@@ -74,6 +74,27 @@
      */
     CompletableFuture<MessageId> sendAsync(T message);
 
+    /**
+     * Flush all the messages buffered in the client when {@link ProducerBuilder#enableBatching(boolean)} is enabled.
+     * It is a no-op if batching is disabled.
```
Comment (merlimat) [inline pulsar-client/src/main/java/org/apache/pulsar/client/api/Producer.java]: The behavior could be consistent even when batching is not being used. 

For example, one could always do: 
```java
producer.sendAsync(msg1);
producer.sendAsync(msg2);
producer.sendAsync(msg3);

producer.flush();
```

This should be equivalent to just waiting on the last future.

Comment (sijie) [reply pulsar-client/src/main/java/org/apache/pulsar/client/api/Producer.java]: sorry the behavior is consistent. it is waiting on the last future. I will improve the documentation.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java [hunk 2]
```diff
@@ -1218,7 +1223,22 @@ public void run(Timeout timeout) throws Exception {
     };
 
     @Override
-    protected void flush() {
+    public CompletableFuture<Void> flushAsync() {
+        if (isBatchMessagingEnabled()) {
```
[the remark this answers is not in the document]
Comment (sijie) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java]: oh i see - I set the lastSendFuture but I forgot to wait :) I will fix it

File: pulsar-broker/src/test/java/org/apache/pulsar/client/api/SimpleProducerConsumerTest.java [hunk 3]
```diff
@@ -2552,4 +2553,74 @@ public void testConsumerSubscriptionInitialize() throws Exception {
...
+        Consumer<byte[]> consumer = pulsarClient.newConsumer().topic("persistent://my-property/my-ns/test-flush-enabled")
+                .subscriptionName("my-subscriber-name").subscribe();
+
+        ProducerBuilder<byte[]> producerBuilder = pulsarClient.newProducer()
+                .topic("persistent://my-property/my-ns/test-flush-enabled")
+                .enableBatching(true)
+                .batchingMaxPublishDelay(1, TimeUnit.HOURS)
+                .batchingMaxMessages(10000);
+
+        @Cleanup
```
Comment (ivankelly) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/api/SimpleProducerConsumerTest.java]: Use try-with-resource here. Achieves same as ```@Cleanup``` and is java standard.

File: pulsar-broker/src/test/java/org/apache/pulsar/client/api/SimpleProducerConsumerTest.java [hunk 4]
```diff
@@ -2552,4 +2553,74 @@ public void testConsumerSubscriptionInitialize() throws Exception {
...
+        consumer.acknowledgeCumulative(msg);
+        consumer.close();
+        log.info("-- Exiting {} test --", methodName);
+    }
+
+    @Test
+    public void testFlushBatchDisabled() throws Exception {
+        log.info("-- Starting {} test --", methodName);
+
+        Consumer<byte[]> consumer = pulsarClient.newConsumer().topic("persistent://my-property/my-ns/test-flush-disabled")
```
Comment (ivankelly) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/api/SimpleProducerConsumerTest.java]: this is longer than 120. I know there's no checkstyle enforced, but we should still try to make stuff readable in github

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java [hunk 5]
```diff
@@ -333,6 +335,7 @@ public void sendAsync(Message<T> message, SendCallback callback) {
                     // batch size and/or max message size
                     if (batchMessageContainer.hasSpaceInBatch(msg)) {
                         batchMessageContainer.add(msg, callback);
+                        lastSendFuture = callback.getFuture();
```
Comment (ivankelly) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java]: Can't we just assign lastSendFuture once, before calling isBatchMessagingEnabled()? Once we get to that point, the message is going to best sent in any case.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java [hunk 6]
```diff
@@ -1218,7 +1223,22 @@ public void run(Timeout timeout) throws Exception {
...
+            CompletableFuture<MessageId> lastSendFuture;
+            synchronized (ProducerImpl.this) {
+                batchMessageAndSend();
+                lastSendFuture = this.lastSendFuture;
+            }
+            if (null != lastSendFuture) {
+                return lastSendFuture.thenApply(ignored -> null);
+            }
+        }
+        return CompletableFuture.completedFuture(null);
```
Comment (ivankelly) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java]: If you initialize lastSendFuture to CompletableFuture.completedFuture(null), you don't need the null check. Also, if batching is not enabled, you should still be waiting on lastSendFuture from the previous asyncSend (not sure if this is what you're referring to in your comment above).

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java [hunk 7]
```diff
@@ -1218,7 +1223,22 @@ public void run(Timeout timeout) throws Exception {
     };
 
     @Override
-    protected void flush() {
+    public CompletableFuture<Void> flushAsync() {
+        if (isBatchMessagingEnabled()) {
+            CompletableFuture<MessageId> lastSendFuture;
+            synchronized (ProducerImpl.this) {
+                batchMessageAndSend();
```
Comment (ivankelly) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java]: batchMessageAndSend is a no-op is batching is disabled, so this whole method could be just
```
public CompletableFuture<Void> flushAsync() {
    synchronized (ProducerImpl.this) {
        batchMessageAndSend();
        return lastSendFuture.thenApply(ignored -> null);
    }
}
```

File: pulsar-broker/src/test/java/org/apache/pulsar/client/api/SimpleProducerConsumerTest.java (hunk 3 above)
Comment (sijie) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/api/SimpleProducerConsumerTest.java]: done

File: pulsar-broker/src/test/java/org/apache/pulsar/client/api/SimpleProducerConsumerTest.java (hunk 4 above)
Comment (sijie) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/api/SimpleProducerConsumerTest.java]: done

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (sijie) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java]: done

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java (hunk 5 above)
Comment (sijie) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java]: fixed

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java (hunk 6 above)
Comment (sijie) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java]: fixed

**Comment by sijie** [comment]
@ivankelly @merlimat addressed your review comments.

**Comment by sijie** [comment]
retest this please

**Pull request merged**